include/tbbexec/tbb_thread_pool.hpp

```diff
@@ -0,0 +1,413 @@
+            // If function invocation doesn't throw
+            stdexec::__nothrow_callable<Fun, Shape, Args...>&&
+            // and emplacing a tuple doesn't throw
+            noexcept(stdexec::__decayed_tuple<Args...>(std::declval<Args>()...))
+            // there's no need to advertise completion with `exception_ptr`
+            >;
+
+        template <class SenderId, class ReceiverId, class Shape, class Fun, bool MayThrow>
+        struct bulk_shared_state : exec::task_base {
+            using Sender = stdexec::__t<SenderId>;
+            using Receiver = stdexec::__t<ReceiverId>;
+
+            using variant_t = stdexec::__value_types_of_t<Sender,
+                                                          stdexec::env_of_t<Receiver>,
+                                                          stdexec::__q<stdexec::__decayed_tuple>,
+                                                          stdexec::__q<stdexec::__variant>>;
+
+            variant_t data_;
+            tbb_thread_pool& pool_;
+            Receiver receiver_;
+            Shape shape_;
+            Fun fn_;
+
+            std::atomic<std::uint32_t> finished_threads_{0};
+            std::atomic<std::uint32_t> thread_with_exception_{0};
+            std::exception_ptr exception_;
+
+            // Splits `n` into `size` chunks distributing `n % size` evenly between ranks.
+            // Returns `[begin, end)` range in `n` for a given `rank`.
+            // Example:
+            // ```cpp
+            // //         n_items  thread  n_threads
+            // even_share(     11,      0,         3); // -> [0,  4) -> 4 items
+            // even_share(     11,      1,         3); // -> [4,  8) -> 4 items
+            // even_share(     11,      2,         3); // -> [8, 11) -> 3 items
+            // ```
+            static std::pair<Shape, Shape> even_share(Shape n, std::uint32_t rank, std::uint32_t size) noexcept {
+                const auto avg_per_thread = n / size;
+                const auto n_big_share = avg_per_thread + 1;
+                const auto big_shares = n % size;
+                const auto is_big_share = rank < big_shares;
+                const auto begin =
+                    is_big_share ? n_big_share * rank : n_big_share * big_shares + (rank - big_shares) * avg_per_thread;
+                const auto end = begin + (is_big_share ? n_big_share : avg_per_thread);
+
+                return std::make_pair(begin, end);
+            }
+
+            std::uint32_t num_agents_required() const {
+                return std::min(shape_, static_cast<Shape>(pool_.available_parallelism()));
+            }
+
+            template <class F>
+            void apply(F f) {
+                std::visit([&](auto& tupl) -> void { std::apply([&](auto&... args) -> void { f(args...); }, tupl); },
+                           data_);
+            }
+
+            bulk_shared_state(tbb_thread_pool& pool, Receiver receiver, Shape shape, Fun fn)
```

Comment:
This is basically all copy-paste from `static_thread_pool.hpp` :-/

Reply:
I don't quite understand what this stuff is trying to do enough to make a suggestion, but it feels like maybe a base class would cover it. So instead of
```c++
template <class SenderId, class ReceiverId, class Shape, class Fun, bool MayThrow>
struct bulk_shared_state : exec::task_base {
    ...
};
```
we'd have
```c++
template <class SenderId, class ReceiverId, class Shape, class Fun, bool MayThrow>
struct bulk_shared_state : exec::task_base_impl<tbb_thread_pool, SenderId, ReceiverId, Shape, Fun, MayThrow> {};
```
which would provide all the goodies.

Reply:
I see where you're coming from, but I would leave it for now until I have a better idea where else we would use such a thing. I'm wary of putting things in the public API until they are more well thought out.

Reply:
I'm happy to leave it for now. (I also could put that in `tbbexec::detail` just so it's factored out.) It does feel like if others are going to write other thread-pool wrappers, it would be more ergonomic for `::exec` to provide some reasonable defaults.